osmoutils/partialord/internal/dag/dag.go

```diff
@@ -0,0 +1,319 @@
+package dag
+
+import (
+	"fmt"
+	"sort"
+)
+
+// Dag struct maintains a directed acyclic graph, using adjacency lists to track edges.
+type Dag struct {
+	// there is a directed edge from u -> v, if directedEdgeList[u][v] = 1
+	// there is a directed edge from v -> u, if directedEdgeList[u][v] = -1
+	directedEdgeList []map[int]int8
+	nodeNameToId     map[string]int
+	idToNodeNames    map[int]string
+}
+
+func NewDag(nodes []string) Dag {
+	nodeNameToId := make(map[string]int, len(nodes))
+	idToNodeNames := make(map[int]string, len(nodes))
+	directedEdgeList := make([]map[int]int8, len(nodes))
+	for i, node := range nodes {
+		nodeNameToId[node] = i
+		idToNodeNames[i] = node
+		directedEdgeList[i] = map[int]int8{}
+	}
+	if len(nodeNameToId) != len(nodes) {
+		panic("provided multiple nodes with the same name")
+	}
+	return Dag{
+		directedEdgeList: directedEdgeList,
+		nodeNameToId:     nodeNameToId,
+		idToNodeNames:    idToNodeNames,
+	}
+}
+
+// Copy returns a new dag struct that is a copy of the original dag.
+// Edges can be mutated in the copy, without altering the original.
+func (dag Dag) Copy() Dag {
+	directedEdgeList := make([]map[int]int8, len(dag.nodeNameToId))
+	for i := 0; i < len(dag.nodeNameToId); i++ {
+		originalEdgeList := dag.directedEdgeList[i]
+		directedEdgeList[i] = make(map[int]int8, len(originalEdgeList))
+		for k, v := range originalEdgeList {
+			directedEdgeList[i][k] = v
+		}
+	}
+	// we re-use nodeNameToId and idToNodeNames as these are fixed at dag creation.
```

Comment:
What's the reason we iterate to copy Dag instead of directly using `dag.directedEdgeList`? Is that to ensure we deep copy edge list?

Reply:
Yeah, the topological sort algorithm used basically removes edges from every node marked as 'done', So we have to remove these edges from a copy.

More space efficient things can technically be done, but should never be important at the relevant sizes here.